src/main/java/org/truffleruby/language/LazyRubyNode.java

```diff
@@ -62,4 +61,25 @@ public RubyNode resolve() {
         });
     }
 
+    private RubyNode getOrCreateMasterResolution() {
+        synchronized (resolver) {
+            RubyNode otherResolution = resolutionMaster[0];
+            RubyNode result;
+            if (otherResolution != null) {
+                result = NodeUtil.cloneNode(otherResolution);
+            } else {
+                if (getContext().getOptions().LAZY_TRANSLATION_LOG) {
+                    Log.LOGGER.info(() -> "lazy translating " + RubyLanguage.fileLine(getParent().getEncapsulatingSourceSection()) + " in " + getRootNode());
+                }
+
+                otherResolution = resolver.get();
+                transferFlagsTo(otherResolution);
+
+                resolutionMaster[0] = otherResolution;
+                result = NodeUtil.cloneNode(otherResolution);
+            }
+            return result;
```

Comment:
Looks great!
I would move the cloneing on this line, and rename `otherResolution` to `masterResolution`.
Then it's clear we always return a copy, and what `masterResolution` holds.

Reply:
Done.